### PR TITLE
All: More info in /journeys on distance and durations

### PR DESCRIPTION
**Pull request by kadhikari**
Attributs bike and car added in journey.durations
Distances with attributs walking, bike and car added in journey
Ticket concerned: https://jira.canaltp.fr/browse/NAVP-704

### CODE REVIEW

**Comment by kadhikari** [comment]
jenkins test this please

File: source/jormungandr/jormungandr/interfaces/v1/serializer/journey.py [hunk 1]
```diff
@@ -100,6 +100,20 @@ class DurationsSerializer(PbNestedSerializer):
...
                                description='Walking total duration of the journey (seconds)')
+    bike = jsonschema.Field(schema_type=int, display_none=True,
+                            description='total duration on bike of the journey (seconds)')
+    car = jsonschema.Field(schema_type=int, display_none=True,
+                           description='total duration in the car of the journey (seconds)')
+
+
+class DistancesSerializer(PbNestedSerializer):
+    walking = jsonschema.Field(schema_type=int, display_none=True,
+                               description='Walking total distance of the journey (seconds)')
```
Comment (TeXitoi) [inline source/jormungandr/jormungandr/interfaces/v1/serializer/journey.py]: Total walking distance of the journey (meters)

File: source/jormungandr/jormungandr/interfaces/v1/serializer/journey.py [hunk 2]
```diff
@@ -100,6 +100,20 @@ class DurationsSerializer(PbNestedSerializer):
                              description='Total duration of the journey (seconds)')
     walking = jsonschema.Field(schema_type=int, display_none=True,
                                description='Walking total duration of the journey (seconds)')
+    bike = jsonschema.Field(schema_type=int, display_none=True,
+                            description='total duration on bike of the journey (seconds)')
```
Comment (TeXitoi) [inline source/jormungandr/jormungandr/interfaces/v1/serializer/journey.py]: Total with a T

File: source/jormungandr/jormungandr/interfaces/v1/serializer/journey.py [hunk 3]
```diff
@@ -100,6 +100,20 @@ class DurationsSerializer(PbNestedSerializer):
...
+    car = jsonschema.Field(schema_type=int, display_none=True,
+                           description='total duration in the car of the journey (seconds)')
```
Comment (TeXitoi) [inline source/jormungandr/jormungandr/interfaces/v1/serializer/journey.py]: Total with a T

File: source/jormungandr/jormungandr/interfaces/v1/serializer/journey.py [hunk 4]
```diff
@@ -100,6 +100,20 @@ class DurationsSerializer(PbNestedSerializer):
...
+
+
+class DistancesSerializer(PbNestedSerializer):
+    walking = jsonschema.Field(schema_type=int, display_none=True,
+                               description='Walking total distance of the journey (seconds)')
+    bike = jsonschema.Field(schema_type=int, display_none=True,
+                            description='total distance on bike of the journey (seconds)')
```
Comment (TeXitoi) [inline source/jormungandr/jormungandr/interfaces/v1/serializer/journey.py]: Total bike distance of the journey (meters)

File: source/jormungandr/jormungandr/interfaces/v1/serializer/journey.py [hunk 5]
```diff
@@ -100,6 +100,20 @@ class DurationsSerializer(PbNestedSerializer):
...
+    car = jsonschema.Field(schema_type=int, display_none=True,
+                           description='total distance in the car of the journey (seconds)')
```
Comment (TeXitoi) [inline source/jormungandr/jormungandr/interfaces/v1/serializer/journey.py]: Total car distance of the journey (meters)

File: source/routing/raptor_api.cpp [hunk 6]
```diff
@@ -187,20 +187,59 @@ static void co2_emission_aggregator(pbnavitia::Journey* pb_journey){
...
+    uint32_t total_car_duration = 0;
+    uint32_t total_bike_duration = 0;
     for (const auto& section: pb_journey->sections()) {
         if ((section.type() == pbnavitia::STREET_NETWORK && section.street_network().mode() == pbnavitia::StreetNetworkMode::Walking) ||
         (section.type() == pbnavitia::TRANSFER && section.transfer_type() == pbnavitia::walking) ||
         (section.type() == pbnavitia::CROW_FLY && section.street_network().mode() == pbnavitia::StreetNetworkMode::Walking)) {
             total_walking_duration += section.duration();
         }
+
+        if (section.type() == pbnavitia::STREET_NETWORK) {
```
Comment (TeXitoi) [inline source/routing/raptor_api.cpp]: CROW_FLY must be handled. Maybe TRANSFER if it can be mutualised, even if we never do something else than walking in transfer for the moment. It should be possible to mutualise with the test for walking.

File: source/routing/raptor_api.cpp [hunk 7]
```diff
@@ -187,20 +187,59 @@ static void co2_emission_aggregator(pbnavitia::Journey* pb_journey){
...
+    uint32_t total_car_distance = 0;
+    uint32_t total_bike_distance = 0;
+    for (const auto& section: pb_journey->sections()) {
+        if ((section.type() == pbnavitia::STREET_NETWORK && section.street_network().mode() == pbnavitia::StreetNetworkMode::Walking) ||
+        (section.type() == pbnavitia::TRANSFER && section.transfer_type() == pbnavitia::walking) ||
+        (section.type() == pbnavitia::CROW_FLY && section.street_network().mode() == pbnavitia::StreetNetworkMode::Walking)) {
+            total_walking_distance += section.length();
+        }
+
+        if (section.type() == pbnavitia::STREET_NETWORK) {
```
Comment (TeXitoi) [inline source/routing/raptor_api.cpp]: Same as durations. All that code is the same as compute_durations, maybe mutualise in a compute_metadata function or something like that.

**Review by TeXitoi — CHANGES_REQUESTED** [review]

File: source/routing/raptor_api.cpp [hunk 8]
```diff
@@ -185,20 +185,46 @@ static void co2_emission_aggregator(pbnavitia::Journey* pb_journey){
...
+                total_walking_distance += section.length();
+            } else if (section.street_network().mode() == pbnavitia::StreetNetworkMode::Car) {
+                total_car_duration += section.duration();
+                total_car_distance += section.length();
+            } else if ((section.street_network().mode() == pbnavitia::StreetNetworkMode::Bike) ||
+                       (section.street_network().mode() == pbnavitia::StreetNetworkMode::Bss)) {
+                total_bike_duration += section.duration();
+                total_bike_distance += section.length();
+            }
+
```
Comment (azime) [inline source/routing/raptor_api.cpp]: minor:
```c++
switch(section.street_network().mode()){
case pbnavitia::StreetNetworkMode::Walking: {
    total_walking_duration += section.duration();
    total_walking_distance += section.length();
    break;
}
case pbnavitia::StreetNetworkMode::Car: {
    total_car_duration += section.duration();
    total_car_distance += section.length();
    break;
}
case pbnavitia::StreetNetworkMode::Bike:
case pbnavitia::StreetNetworkMode::Bss: {
    total_bike_duration += section.duration();
    total_bike_distance += section.length();
    break;
}
}
```

File: source/jormungandr/jormungandr/interfaces/v1/serializer/journey.py [hunk 9]
```diff
@@ -100,6 +100,20 @@ class DurationsSerializer(PbNestedSerializer):
                              description='Total duration of the journey (seconds)')
     walking = jsonschema.Field(schema_type=int, display_none=True,
                                description='Walking total duration of the journey (seconds)')
+    bike = jsonschema.Field(schema_type=int, display_none=True,
+                            description='Total duration on bike of the journey (seconds)')
+    car = jsonschema.Field(schema_type=int, display_none=True,
+                           description='Total car duration of the journey (meters)')
```
Comment (pbougue) [inline source/jormungandr/jormungandr/interfaces/v1/serializer/journey.py]: Uncorrelated description.

File: source/jormungandr/jormungandr/interfaces/v1/serializer/journey.py [hunk 10]
```diff
@@ -100,6 +100,20 @@ class DurationsSerializer(PbNestedSerializer):
...
+
+
+class DistancesSerializer(PbNestedSerializer):
+    walking = jsonschema.Field(schema_type=int, display_none=True,
+                               description='Total walking distance of the journey (meters)')
+    bike = jsonschema.Field(schema_type=int, display_none=True,
+                            description='Total bike distance of the journey (meters)')
+    car = jsonschema.Field(schema_type=int, display_none=True,
+                           description='Total walking distance in the car of the journey (seconds)')
```
Comment (pbougue) [inline source/jormungandr/jormungandr/interfaces/v1/serializer/journey.py]: Uncorrelated description.

File: source/jormungandr/tests/journey_common_tests.py [hunk 11]
```diff
@@ -525,6 +527,17 @@ def test_max_duration_to_pt_equals_to_0(self):
         self.is_valid_journey_response(response, query)
         assert len(response['journeys']) == 4
 
+        assert response['journeys'][0]['durations']['bike'] == 62
+        assert response['journeys'][0]['durations']['total'] == 62
+        assert response['journeys'][0]['distances']['bike'] == 257
+        assert response['journeys'][0]['durations']['walking'] == 0
```
Comment (pbougue) [inline source/jormungandr/tests/journey_common_tests.py:532]: Minor (readability) : to be moved above

File: source/routing/raptor_api.cpp [hunk 12]
```diff
@@ -185,20 +185,46 @@ static void co2_emission_aggregator(pbnavitia::Journey* pb_journey){
...
+        } else if (section.type() == pbnavitia::TRANSFER && section.transfer_type() == pbnavitia::walking) {
```
Comment (pbougue) [inline source/routing/raptor_api.cpp]: Minor : maybe also test that transfer_type exists (if None, then it's walking)

**Review by pbougue — APPROVED** [review]
We should add those fields to artemis

File: source/routing/raptor_api.cpp (hunk 12 above)
Comment (TeXitoi) [reply source/routing/raptor_api.cpp]: It can be a stay_in, thus the test is needed

File: source/routing/tests/routing_api_test.cpp [hunk 13]
```diff
@@ -1040,6 +1040,7 @@ BOOST_FIXTURE_TEST_CASE(biking, streetnetworkmode_fixture<test_speed_provider>)
     BOOST_CHECK_EQUAL(resp.journeys(1).has_co2_emission(), true);
     BOOST_CHECK_EQUAL(resp.journeys(1).co2_emission().value(), 0.);
     BOOST_CHECK_EQUAL(resp.journeys(1).co2_emission().unit(), "gEC");
+
```
Comment (xlqian) [inline source/routing/tests/routing_api_test.cpp]: useless blank line

File: source/routing/tests/routing_api_test.cpp [hunk 14]
```diff
@@ -1069,6 +1070,14 @@ BOOST_FIXTURE_TEST_CASE(biking_walking, streetnetworkmode_fixture<test_speed_pro
...
     BOOST_CHECK_EQUAL(section.street_network().duration(), 130); //it's the biking distance / biking speed (but there can be rounding pb)
     BOOST_REQUIRE_EQUAL(section.street_network().path_items_size(), 7);
+    BOOST_CHECK_EQUAL(journey.durations().walking(), 0);
+    BOOST_CHECK_EQUAL(journey.durations().bike(), 130);
+    BOOST_CHECK_EQUAL(journey.durations().car(), 0);
+    BOOST_CHECK_EQUAL(journey.durations().total(), 130);
+    BOOST_CHECK_EQUAL(journey.distances().walking(), 0);
+    BOOST_CHECK_EQUAL(journey.distances().bike(), 532);
+    BOOST_CHECK_EQUAL(journey.distances().car(), 0);
+
```
Comment (xlqian) [inline source/routing/tests/routing_api_test.cpp]: useless blank line

**Review by xlqian — APPROVED** [review]

File: source/jormungandr/jormungandr/interfaces/v1/serializer/journey.py [hunk 15]
```diff
@@ -100,6 +100,20 @@ class DurationsSerializer(PbNestedSerializer):
                              description='Total duration of the journey (seconds)')
     walking = jsonschema.Field(schema_type=int, display_none=True,
                                description='Walking total duration of the journey (seconds)')
```
Comment (TeXitoi) [inline source/jormungandr/jormungandr/interfaces/v1/serializer/journey.py]: Total walking duration of the journey (seconds)

File: source/jormungandr/jormungandr/interfaces/v1/serializer/journey.py [hunk 16]
```diff
@@ -100,6 +100,20 @@ class DurationsSerializer(PbNestedSerializer):
...
+    bike = jsonschema.Field(schema_type=int, display_none=True,
+                            description='Total duration on bike of the journey (seconds)')
```
Comment (TeXitoi) [inline source/jormungandr/jormungandr/interfaces/v1/serializer/journey.py]: Total bike duration of the journey (seconds)

File: source/jormungandr/jormungandr/interfaces/v1/serializer/journey.py [hunk 17]
```diff
@@ -100,6 +100,20 @@ class DurationsSerializer(PbNestedSerializer):
...
+                           description='Total car duration of the journey (seconds)')
+
+
+class DistancesSerializer(PbNestedSerializer):
+    walking = jsonschema.Field(schema_type=int, display_none=True,
+                               description='Total walking distance of the journey (meters)')
+    bike = jsonschema.Field(schema_type=int, display_none=True,
+                            description='Total bike distance of the journey (meters)')
+    car = jsonschema.Field(schema_type=int, display_none=True,
+                           description='Total walking distance in the car of the journey (meters)')
```
Comment (TeXitoi) [inline source/jormungandr/jormungandr/interfaces/v1/serializer/journey.py]: Total car distance of the journey (meters)

**Review by TeXitoi — CHANGES_REQUESTED** [review]

**Review by TeXitoi — APPROVED** [review]
